pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol

```diff
@@ -921,6 +937,111 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
         }
     }
 
+    // Circuit Breakers
+
+    /**
+     * @notice Return the circuit breaker parameters for the given token.
+     * @dev These are the reference values (BPT price and weight complement) computed when the breaker was set,
+     * along with the percentage bounds. To get the current BPT price boundaries needed to check whether the
+     * circuit breaker should trip, call `getCurrentCircuitBreakerBounds`.
+     */
+    function getCircuitBreakerFields(IERC20 token)
+        external
+        view
+        returns (
+            uint256,
+            uint256,
+            uint256,
+            uint256
+        )
+    {
+        return CircuitBreakerLib.getCircuitBreakerFields(_circuitBreakerState[token]);
+    }
+
+    /**
+     * @notice Get the current BPT price bounds for the given token.
+     * @dev The bounds are set as percentages, but retrieved as BPT prices which can be directly compared to the
+     * current BPT price. Though the percentages are fixed when the breaker is set, the BPT price bounds are dynamic.
+     * If there is an ongoing gradual weight update, they will shift up or down with the current weight.
+     */
+    function getCurrentCircuitBreakerBounds(IERC20 token) public view returns (uint256, uint256) {
+        return
+            CircuitBreakerLib.getCurrentCircuitBreakerBounds(
+                _circuitBreakerState[token],
+                _getNormalizedWeight(token).complement()
+            );
+    }
+
+    /**
+     * @notice Returns flags indicating whether the lower or upper bound circuit breakers have tripped.
+     */
+    function hasCircuitBreakerTripped(IERC20 token) external view returns (bool, bool) {
+        return
+            CircuitBreakerLib.hasCircuitBreakerTripped(
+                _circuitBreakerState[token],
+                getActualSupply(),
+                _getNormalizedWeight(token),
+                _getUpscaledTokenBalance(token)
+            );
+    }
+
+    /**
+     * @notice Set a circuit breaker for a token.
+     * @dev This is a permissioned function, and disabled if the pool is paused. The lower and upper bounds
+     * are percentages, corresponding to a *relative* change in the token's spot price: e.g., a lower bound
+     * of 0.8 means the breaker should prevent trades that result in the value of the token dropping 20% or
+     * more relative to the rest of the pool.
+     */
+    function setCircuitBreaker(
+        IERC20 token,
+        uint256 lowerBoundPercentage,
+        uint256 upperBoundPercentage
+    ) external authenticate whenNotPaused {
+        _setCircuitBreaker(token, lowerBoundPercentage, upperBoundPercentage);
+    }
```

Comment:
I've got some questions about whether this should accept an array of inputs. This would be consistent with `updateWeightsGradually` and avoid the manager having to call in many times.

Reply:
We do have a requirement to be able to set them individually, but there's no requirement here that the token list match the Vault tokens (e.g., you could pass in an array of 1), so that should be fine.

Reply:
Yeah, that's what I was picturing. Sorry the comparison to `updateWeightsGradually` was a bit confusing.